src/core_plugins/kibana/public/dashboard/reducers/panels.js

```diff
@@ -0,0 +1,44 @@
+import _ from 'lodash';
+import { handleActions } from 'redux-actions';
+
+import {
+  deletePanel,
+  updatePanel,
+  updatePanels,
+  embeddableRenderFinished,
+  embeddableRenderError
+} from '../actions';
+
+export const panels = handleActions({
+  [deletePanel]: (state, { payload }) => {
+    const stateCopy = { ...state };
+    delete stateCopy[payload];
+    return stateCopy;
+  },
+
+  [updatePanel]: (state, { payload }) => ({
+    ...state,
+    [payload.panelIndex]: _.defaultsDeep(payload, state[payload.panelIndex])
+  }),
+
+  [updatePanels]: (state, { payload }) => ({
+    ..._.cloneDeep(payload)
+  }),
+
+  [embeddableRenderFinished]: (state, { payload }) => ({
+    ...state,
+    [payload.panelId]: {
+      ...state[payload.panelId],
+      renderError: null,
+      embeddable: payload.embeddable,
+    }
+  }),
+
+  [embeddableRenderError]: (state, { payload: { panelId, error } }) => ({
+    ...state,
+    [panelId]: {
```

Comment:
I had good experience with this pattern: https://github.com/elastic/cloud/blob/master/cloud-ui/public/reducers/allocators/index.js#L20-L32

Basically creating a reducer for each item (here: panel), in addition to having the top one for `panels`. That made growing the reducers over time easier in my experience.